### PR TITLE
Replace tab completion generation system

### PR DESCRIPTION
**Pull request by georgebrock**
This commit replaces the `Gitsh::Completer` with a more sophisticated system. For an overview of how the new system works, see `lib/gitsh/tab_completion/README.md`, which is added by this commit.

The new system has two benefits over the old:

1. It makes use of the `Gitsh::Lexer` to split up the command, instead of naively splitting the input on spaces. This means we're able to tab complete commands that don't appear at the beginning of the line, e.g. when entering `add . && com<tab>` or `:echo $(con<tab>`.

2. It uses a non-deterministic finite automaton to understand the context of the completion, instead of just distinguishing between commands and arguments. While the NFA in this commit doesn't do any more than the old completer did, it lays the ground work for much more sophisticated context-aware completions in the future.

---

Pair: @tabfugnic

### CODE REVIEW

**Comment by georgebrock** [comment]
This fixes #261, and takes us a big step closer to #25, #80, and #220.

File: lib/gitsh/tab_completion/context.rb [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+      def prior_words
+        words[0...-1]
+      end
+
+      private
+
+      attr_reader :input
+
+      def words
+        @_words ||= tokens.reverse_each.inject(['']) do |words, token|
```
Comment (mike-burns) [inline lib/gitsh/tab_completion/context.rb]: How about:

```ruby
def words
  combine_words(split_by_spaces(last_command))
end

def last_command
  tokens.reverse_each.take_while do |token|
    [:AND, :OR, :SEMICOLON, :LEFT_PAREN, :SUBSHELL_START, :EOL].exclude?(token.type)
  end.reverse
end

# split_by_spaces : [token] -> [[token]]
def split_by_spaces
  chunk { |token| token.type == :SPACE }.
    inject([]) do |result, (is_space, tokens)|
      if is_space
        result
      else
        result + [tokens]
      end
    end
end

# combine_words : [[word tokens]] -> [string]
def combine_words
  map do |tokens|
    tokens.select { |token| token.type == :WORD }.map(&:value).join
  end
end
```

[From pairing with @georgebrock @tabfugnic .]

File: lib/gitsh/tab_completion/automaton.rb [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+      def match(tokens)
+        tokens.inject(start_states) do |current_states, token|
+          current_states.map { |state| state.follow(token) }.inject(Set.new, :|)
+        end
+      end
+
+      def completions(context, token)
+        match(context).
+          flat_map { |state| state.completions(token) }.
+          select { |option| option.start_with?(token) }
```
Comment (mike-burns) [inline lib/gitsh/tab_completion/automaton.rb]: How about moving this to `BaseMatcher`, to make this automaton definition more clear?

File: lib/gitsh/tab_completion/context.rb [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+      attr_reader :input
+
+      def words
+        combine_words(split_by_spaces(last_command(tokens)))
+      end
+
+      def combine_words(token_groups)
+        token_groups.map do |tokens|
+          tokens.select { |token| token.type == :WORD }.map(&:value).join
```
Comment (mike-burns) [inline lib/gitsh/tab_completion/context.rb]: Could turn these three into an `#inject`:

```ruby
tokens.inject([]) do |result, token|
  if token.type == :WORD
    "#{result}#{token.value}"
  else
    result
  end
end
```

File: lib/gitsh/tab_completion/matchers/path_matcher.rb [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+            File.expand_path(token) + '/'
+          else
+            File.expand_path(token)
+          end
+        end
+
+        def paths(prefix)
+          Dir["#{prefix}*"].map do |path|
+            if File.directory?(path)
+              path + '/'
```
Comment (mike-burns) [inline lib/gitsh/tab_completion/matchers/path_matcher.rb]: Is this too much abstraction?

```ruby
def normalize_path(token)
  conditionally_slash(File.expand_path(token)) { token.end_with?('/') }
end

def paths(prefix)
  Dir["#{prefix}*"].map do |path|
    conditionally_slash(path) { File.directory?(path) }
  end
end

def conditionally_slash(string, &predicate)
  if predicate.call
    string + '/'
  else
    string
  end
end
```

File: spec/units/tab_completion/matchers/text_matcher_spec.rb [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+      expect(matcher.completions('commit')).to eq ['commit']
+    end
+  end
+
+  describe '#eql?' do
+    it 'returns true for text matchers with the same text' do
+      x1 = described_class.new('x')
+      x2 = described_class.new('x')
+
+      expect(x1.eql?(x2)).to be true
```
Comment (mike-burns) [inline spec/units/tab_completion/matchers/text_matcher_spec.rb]: Any reason to write it like this instead of `expect(x1).to eql(x2)`?

**Comment by mike-burns** [comment]
BTW, I had some code golf ideas, but it LGTM at any point.

File: lib/gitsh/tab_completion/matchers/path_matcher.rb (hunk 4 above)
Comment (georgebrock) [inline lib/gitsh/tab_completion/matchers/path_matcher.rb]: I think it is a bit too much: if I saw this in practice, I'd need to read the `conditionally_slash` method to understand what was going on, so I'm going to leave it inline.

**Pull request merged**